README.md
# SimpleSDP
SimpleSDP is a laravel package that makes it easy to send SMS and OTP(One Time Password) via SDP.


> Note: farsi [documentation](README.fa.md)

## Some of the features for SimpleSDP:
- Support multiple drivers
- Configurable
- Customize models
- Logging system
- Translatable
- Auto discovery (only laravel 5.5+)

## Requirements:
- Laravel 5.3 or higher
- PHP 5.6.4 or higher

## Available Drivers
- AKO (ako)
- SSDP (ssdp)

## Installation

1. ##### Install the package via composer:
    ```shell
    composer require mostafaznv/simple-sdp
    ```

2. ##### Register Provider and Facade in config/app.php (Required for Laravel 5.3 and 5.4):
    ```shell
    'providers' => [
      ...
      Mostafaznv\SimpleSDP\SimpleSDPServiceProvider::class,
    ],
    
    
    'aliases' => [
      ...
      'SimpleSDP' => Mostafaznv\SimpleSDP\SimpleSDP::class,
    ]
    ```

3. ##### Publish config, translation and migrations:
    ```shell
    php artisan vendor:publish --provider="Mostafaznv\SimpleSDP\SimpleSDPServiceProvider"
    ```

4. ##### Create tables:
    ```shell
    php artisan migrate
    ```

4. ##### Done

> to set your own config, just open config/simple-sdp.php and set your personal configurations.

## Usage
```php
try {   
   $msisdn = '9891200012345';
   $data = [
       'content_id' => 12,
       'message' => 'message text'
   ];   
   
   $sdp = app('SimpleSDP')->sendMt($msisdn, $data); // use simple sdp instance with app() method.

   // or

   $sdp = \SimpleSDP::sendMt($msisdn, $data); // send mt to default driver (simple-sdp can load default driver from config file).
   
   // or

   $sdp = \SimpleSDP::AKO()->sendMt($msisdn, $data); // define driver manually
   
   // or
   
   $sdp = \SimpleSDP::make(new \Mostafaznv\SimpleSDP\SSDP\SSDP())->sendMt($msisdn, $data); // define driver manually
   
   return response()->json($sdp, 200);
  
} 
catch (\Exception $e) {   
    return response()->json($e->getMessage(), 500);
}
```

## Dynamic Configuration
You can pass a new config array to SimpleSDP instance to replace with default config array.
```php
// default config
$sdp = app('SimpleSDP')->AKO()->sendMt($msisdn, $data); 

// new config
$config = config('simple-sdp.new_config');
$sdp = app('SimpleSDP')->AKO($config)->sendMt($msisdn, $data);

```

## Custom Trans prefix (only available in ako)
Sometimes you want change messages in some cases based on your project logic, with this option you can do it.

1. Change config file
    ```php
    ...
    'trans_prefix' => 'new_prefix'
    ```
2. Edit messages array in `resources/lang/vendor/simple-sdp/en/messages.php`
    ```php
    ...
    'new_prefix' => [
       'welcome'     => 'welcome',
       'inform'      => 'inform',
       'guide'       => 'guide',
       'unsub-guide' => 'unsub guide',
       'unsub'       => 'unsub',
       'logout_text' => 'logout text',
    ]
    ```

## Available Functions
1. #### Send MT
    ```php
    $msisdn = '9891200012345';
    $data = [
        'content_id' => 12,
        'message' => 'message text',
        'is_free' => true', // optional
    ];
    
    app('SimpleSDP')->sendMt($msisdn, $data);
    ```
2. #### Send Batch MT
    ```php
    $msisdn = ['9891200012345', '9891200015432'];
    $data = [
        'content_id' => 12,
        'message' => 'message text'
    ];
    
    app('SimpleSDP')->sendBatchMt($msisdn, $data);
    ```
    
3. #### Charge
    ```php
    $msisdn = '9891200012345';
    $data = [
        'content_id' => 12
    ];
    
    app('SimpleSDP')->charge($msisdn, $data);
    ```
    
4. #### Send OTP
    ```php
    $msisdn = '9891200012345';
    $data = [
        'content_id' => 12
    ];
    
    app('SimpleSDP')->sendOtp($msisdn, $data);
    ```
    
5. #### Confirm OTP
    ```php
    $msisdn = '9891200012345';
    $code = 2323;
    $data = [
        'content_id' => 12
    ];
    
    app('SimpleSDP')->confirmOtp($msisdn, $code, $data);
    ```
    
6. #### Delivery
    ```php
    app('SimpleSDP')->delivery($request); // Request $request
    ```
    
7. #### Batch Delivery
    ```php
    app('SimpleSDP')->batchDelivery($request); // Request $request
    ```
    
8. #### Incoming Message
    ```php
    app('SimpleSDP')->income($request); // Request $request
    ```
    
9. #### Batch Mo
    ```php
    app('SimpleSDP')->batchMo($request); // Request $request
    ```
    

## Contributors
- Mostafa Zeinivand [@mostafaznv](https://github.com/mostafaznv)
- Faezeh Ghorbannezhad [@Ghorbannezhad](https://github.com/Ghorbannezhad)
- SamssonApps [@SamssonApps](https://github.com/SamssonApps)


## Changelog
Refer to the [Changelog](CHANGELOG.md) for a full history of the project.

## License
This software is released under [The MIT License (MIT)](LICENSE).

(c) 2018 Mostafaznv, All rights reserved.

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

## v0.0.4 - 2019/05/06
- Rashin Driver
- debug AKO

## v0.0.2 - 2018/10/08
- support wap
- add driver name to all tables
- custom trans messages
- custom configuration
- debug


## v0.0.1 - 2018/10/06
- Initial Release


src/AKO/AKO.php
<?php

namespace Mostafaznv\SimpleSDP\AKO;

use GuzzleHttp\Client as Guzzle;
use GuzzleHttp\Exception\GuzzleException;
use Illuminate\Validation\Rule;
use Mostafaznv\SimpleSDP\Enum;
use Mostafaznv\SimpleSDP\SdpAbstract;
use Mostafaznv\SimpleSDP\SdpInterface;
use Illuminate\Http\Request;
use Carbon\Carbon;

class AKO extends SdpAbstract implements SdpInterface
{
    /**
     * Load MobileTerminated model dynamically.
     *
     * @var \Mostafaznv\SimpleSDP\Models\MobileTerminated
     */
    protected $mobileTerminated;

    /**
     * Load MobileOriginated model dynamically.
     *
     * @var \Mostafaznv\SimpleSDP\Models\MobileOriginated
     */
    protected $mobileOriginated;

    public function boot()
    {
        parent::boot();

        $this->mobileTerminated = config('simple-sdp.models.mobile_terminated');
        $this->mobileOriginated = config('simple-sdp.models.mobile_originated');
    }

    public function sendMt($msisdn, Array $data)
    {
        $this->validate($data, ['content_id' => 'required', 'message' => 'required', 'is_free' => 'nullable|boolean']);

        $this->log('start send mt to ' . $msisdn);

        $url = $this->url('sendmessage/');
        $description = strtr($this->config['description'], [
            ':short_code' => $this->config['short_code'],
            ':content_id' => $data['content_id']
        ]);

        $requestBody = [
            'username'    => $this->config['username'],
            'password'    => $this->config['password'],
            'serviceid'   => $this->config['service_id'],
            'shortcode'   => $this->config['short_code'],
            'msisdn'      => $msisdn,
            'description' => $description,
            'chargecode'  => $this->config['sub_charging_code'],
            'amount'      => $this->config['amount'],
            'currency'    => $this->config['currency'],
            'message'     => $data['message'],
            'is_free'     => isset($data['is_free']) ? $data['is_free'] : $this->config['is_free'],
            'correlator'  => $this->config['short_code'] . '-' . $this->uniqid(),
            'servicename' => $this->config['service_name'],
        ];

        $this->log(json_encode($requestBody));

        $client = new Guzzle();
        $response = $client->request('POST', $url, ['form_params' => $requestBody]);
        $result = json_decode($response->getBody());

        if ($response->getStatusCode() == 200) {
            if ($result->status_code == 0) {
                $mt = new $this->mobileTerminated;
                $connection = (!empty($this->config['database']))?$this->config['database']:null;
                if($connection){
                    $mt->setConnection($connection);
                }
                $mt->msisdn = $msisdn;
                $mt->message = $data['message'];
                $mt->message_id = $requestBody['correlator'];
                $mt->type = (isset($data['type']))?$data['type']:$this->mobileTerminated::DEFAULT_TYPE;
                $mt->driver = $this->driverName;
                $mt->creator_ip = $this->request->ip();
                $mt->save();

                $this->log('done');

                return $this->response(true, Enum::SUCCESS_CODE);
            }
        }

        $this->log(json_encode($result), 'error');

        return $this->response(false, Enum::UNKNOWN_ERROR_CODE);
    }

    public function sendBatchMt(array $msisdn, array $data)
    {
        $this->validate($data, ['message' => 'required', 'content_id' => 'required', 'is_free' => 'nullable|boolean']);

        $this->log('start send batch mt to ' . count($msisdn) . ' users');

        $url = $this->url('bsendmessage/');
        $description = strtr($this->config['description'], [
            ':short_code' => $this->config['short_code'],
            ':content_id' => $data['content_id']
        ]);

        $list = [];
        foreach ($msisdn as $value) {
            $list[] = [
                'msisdn'     => $value,
                'message'    => $data['message'],
                'correlator' => $this->config['short_code'] . '-' . $this->uniqid(),
            ];
        }

        $this->log($url);

        $requestBody = [
            'username'    => $this->config['username'],
            'password'    => $this->config['password'],
            'serviceid'   => $this->config['service_id'],
            'shortcode'   => $this->config['short_code'],
            'description' => $description,
            'chargecode'  => $this->config['sub_charging_code'],
            'amount'      => $this->config['amount'],
            'currency'    => $this->config['currency'],
            'is_free'     => isset($data['is_free']) ? $data['is_free'] : $this->config['is_free'],
            'servicename' => $this->config['service_name'],
            'list'        => $list
        ];

        $this->log(json_encode($requestBody));

        try {
            $client = new Guzzle();
            $response = $client->request('post', $url, ['json' => $requestBody]);
            $result = json_decode($response->getBody());

            if ($response->getStatusCode() == 200) {
                if ($result->status_code == 0) {
                    $insertData = [];

                    foreach ($list as $value) {
                        $insertData[] = [
                            'msisdn'     => $value['msisdn'],
                            'message'    => $value['message'],
                            'message_id' => $value['correlator'],
                            'type'       => $this->mobileTerminated::DEFAULT_TYPE,
                            'driver'     => $this->driverName,
                            'creator_ip' => $this->request->ip(),
                        ];
                    }

                    $this->mobileTerminated::insert($insertData);

                    $this->log('done');

                    return $this->response(true, Enum::SUCCESS_CODE);
                }

                $this->log(json_encode($result), 'error');
            }
        }
        catch (GuzzleException $e) {
            $this->log($e->getMessage(), 'error');
        }

        return $this->response(false, Enum::UNKNOWN_ERROR_CODE);
    }

    public function charge($msisdn, Array $data)
    {
        $this->validate($data, ['content_id' => 'required']);

        $this->log('start charge to ' . $msisdn);

        $url = $this->url('charging/');
        $description = strtr($this->config['description'], [
            ':short_code' => $this->config['short_code'],
            ':content_id' => $data['content_id']
        ]);


        $requestBody = [
            'username'    => $this->config['username'],
            'password'    => $this->config['password'],
            'serviceid'   => $this->config['service_id'],
            'msisdn'      => $msisdn,
            'description' => $description,
            'chargecode'  => $this->config['charging_code'],
            'amount'      => $this->config['amount'],
            'currency'    => $this->config['currency'],
            'correlator'  => $this->config['charging_code'] . '-' . $this->uniqid(),
        ];

        $this->log(json_encode($requestBody));

        $client = new Guzzle();
        $response = $client->request('POST', $url, ['form_params' => $requestBody]);
        $result = json_decode($response->getBody());

        if ($response->getStatusCode() == 200) {
            if ($result->status_code == 0) {
                $mt = new $this->mobileTerminated;
                $connection = (!empty($this->config['database']))?$this->config['database']:null;
                if($connection){
                    $mt->setConnection($connection);
                }
                $mt->msisdn = $msisdn;
                $mt->message = '';
                $mt->message_id = $requestBody['correlator'];
                $mt->type = $this->mobileTerminated::CHARGE_TYPE;
                $mt->driver = $this->driverName;
                $mt->creator_ip = $this->request->ip();
                $mt->save();

                $this->log('done');

                return $this->response(true, Enum::SUCCESS_CODE);
            }
        }

        $this->log(json_encode($result), 'error');

        return $this->response(false, Enum::UNKNOWN_ERROR_CODE);
    }

    public function sendOtp($msisdn, Array $data)
    {
        $this->validate($data, ['content_id' => 'required']);

        $this->log('start send otp to ' . $msisdn);

        $url = $this->url('otp-generation/');
        $requestBody = [
            'msisdn'        => $msisdn,
            'username'      => $this->config['username'],
            'password'      => $this->config['password'],
            'servicename'   => $this->config['service_name'],
            'serviceid'     => $this->config['service_id'],
            'referencecode' => $this->config['short_code'] . '-' . $this->uniqid(),
            'shortcode'     => $this->config['short_code'],
            'contentid'     => $data['content_id'],
            'chargecode'    => $this->config['sub_charging_code'],
            'description'   => 'OTP',
            'amount'        => $this->config['amount'],
        ];

        $this->log(json_encode($requestBody));

        $client = new Guzzle();
        $response = $client->request('POST', $url, ['form_params' => $requestBody]);

        if ($response->getStatusCode() == 200) {
            $result = json_decode($response->getBody());

            $statusCode = $result->data->statusInfo->statusCode;
            if ($statusCode == 200) {
                $mt = new $this->mobileTerminated;
                $connection = (!empty($this->config['database']))?$this->config['database']:null;
                if($connection){
                    $mt = $mt->setConnection($connection);
                }
                $mt->msisdn = $msisdn;
                $mt->message = $data['content_id'];
                $mt->message_id = $requestBody['referencecode'];
                $mt->transaction_id = $result->data->statusInfo->OTPTransactionId;
                $mt->type = isset($data['type']) ? $data['type'] : $this->mobileTerminated::OTP_TYPE;
                $mt->driver = $this->driverName;
                $mt->creator_ip = $this->request->ip();
                $mt->save();

                $this->log('done');

                return $this->response(true, Enum::SUCCESS_CODE);
            }
            elseif ($statusCode == 500) {
                $error = $result->data->statusInfo->errorInfo->errorCode;
                $code = Enum::UNKNOWN_ERROR_CODE;

                if ($error == 51035) {
                    $code = Enum::ALREADY_EXISTS_CODE;
                    $this->log('already exists', 'error');
                }
                else {
                    $this->log('unknown error', 'error');
                }

                return $this->response(false, $code);
            }
        }

        return $this->response(false, Enum::UNKNOWN_ERROR_CODE);
    }

    public function confirmOtp($msisdn, $code, Array $data)
    {
        $this->validate($data, ['content_id' => 'required']);

        $this->log("confirm otp started for $msisdn with code $code");

        $timeout = Carbon::now()->subMinute($this->config['confirm_otp_timeout']);
        $type = (isset($data['type']))?$data['type']:$this->mobileTerminated::OTP_TYPE;

        $mt = $this->mobileTerminated::where('msisdn', $msisdn)->where('type', $type)->whereDate('created_at', '<=', $timeout)->orderby('id', 'desc');
        $connection = (!empty($this->config['database']))?$this->config['database']:null;

        if($connection){
            $this->log('connection is not empty'.$connection);
            $mt->on($connection);
        }

        $mt = $mt->first();

        if ($mt) {
            $url = $this->url('otp-confirmation/');
            $data = [
                'msisdn'         => $msisdn,
                'username'       => $this->config['username'],
                'password'       => $this->config['password'],
                'serviceid'      => $this->config['service_id'],
                'referencecode'  => $mt->message_id,
                'shortcode'      => $this->config['short_code'],
                'contentid'      => $data['content_id'],
                'message'        => $code,
                'otptransaction' => $mt->transaction_id,
            ];

            $client = new Guzzle();
            try{
                $response = $client->request('POST', $url, ['form_params' => $data]);

                if ($response->getStatusCode() == 200) {
                    $result = json_decode($response->getBody());
                    $this->log($response->getBody());
                    $status_code = $result->data->statusInfo->statusCode;
                    if ($status_code == 200) {
                        $versionName = $this->request->version_name ? $this->request->version_name : null;
                        $ip = $this->request->ip();
                        $this->log("type of confirm log: $type");
                        $this->logConfirmOtp($msisdn,$type, $versionName, $ip);

                        $this->log('done');
                        return $this->response(true, Enum::SUCCESS_CODE);
                    }
                }

                $this->log('unknown error', 'error');
                return $this->response(false, Enum::UNKNOWN_ERROR_CODE);
            }catch (\Exception $e){
                $this->log('Exception', 'error');
                return $this->response(false, Enum::UNKNOWN_ERROR_CODE);
            }

        }
        else {
            $this->log('mt not found', 'error');
            return $this->response(false, Enum::MT_NOT_FOUND_CODE);
        }
    }

    public function delivery(Request $request)
    {
        $this->log('delivery received');

        $validator = $this->validate($request->all(), [
            'msisdn'         => ['required','regex:/^tel:989[0-9]{9}$/'],
            'correlator'     => 'required',
            'deliverystatus' => 'required'
        ], false);

        if ($validator['status']) {
            $this->log('for ' . $request->msisdn);
            $msisdn = trim(str_replace('tel:','',$request->msisdn));
            $mt = $this->mobileTerminated::where('msisdn', $msisdn)->where('message_id', $request->correlator)->first();

            if ($mt) {
                $mt->delivery_status = $request->deliverystatus;
                $mt->updater_ip = $request->ip();
                $mt->save();

                $this->log('done');

                return $this->response(true, Enum::SUCCESS_CODE);
            }

            $this->log('unknown error', 'error');
            return $this->response(false, Enum::UNKNOWN_ERROR_CODE);
        }
        else {
            $this->log('not valid input ' . json_encode($validator['errors']), 'error');
            return $this->response(false, Enum::NOT_VALID_INPUTS_CODE);
        }
    }

    public function batchDelivery(Request $request)
    {
        $this->log('batch delivery called');

        $xml = $request->getContent();

        $cleanXml = str_ireplace(['soapenv:', 'loc:'], '', $xml);
        $xml = simplexml_load_string($cleanXml);
        $collection = (array)$xml->Body->notifySmsDeliveryReceipt->notifySmsDeliveryReceiptCollection;

        $result = [];
        foreach ((array)$collection['NotifySmsDeliveryReceipt'] as $item) {
            $item = (array)$item;
            $itemDeliveryStatus = (array)$item['deliveryStatus'];
            $address = str_replace('tel:', '', $itemDeliveryStatus['address']);

            $result[] = $this->handleBatchDelivery((string)$address, (string)$item["correlator"], (string)$itemDeliveryStatus['deliveryStatus'], $request->ip());
        }

        if (!in_array(false, $result)) {
            $this->log('done');
            $this->log($result);

            return $this->response(true, Enum::SUCCESS_CODE, null, $result);
        }
        else {
            $this->log('failed', 'error');
            $this->log($result, 'error');

            return $this->response(false, Enum::UNKNOWN_ERROR_CODE, null, $result);
        }
    }

    private function handleBatchDelivery($msisdn, $correlator, $deliveryStatus, $ip = '0:0:0:0')
    {
        $mt = $this->mobileTerminated::where('msisdn', $msisdn)->where('message_id', $correlator)->first();
        if ($mt) {
            $mt->delivery_status = $deliveryStatus;
            $mt->updater_ip = $ip;
            if ($mt->save()) {
                return true;
            }
        }

        return false;
    }

    public function income(Request $request)
    {
        $this->log('income message received');
        $validator = $this->validate($request->all(), [
            'msisdn'      => ['required', 'digits:12', 'regex:/^989[0-9]{9}$/'],
            'shortcode'   => [
                'required',
                'numeric',
                Rule::in([$this->config['short_code']])
            ],
            'message'     => 'required|string',
            /* 'partnername' => 'string',
             'trans_id'    => 'string',*/
        ], false);

        if ($validator['status']) {
            return $this->handleIncome($request->msisdn, $request->shortcode, $request->trans_id, $request->message, $request->ip());
        }
        else {
            $this->log('not valid input ' . json_encode($validator['errors']), 'error');
            return $this->response(false, Enum::NOT_VALID_INPUTS_CODE);
        }
    }

    private function handleIncome($msisdn, $shortcode, $transactionId, $message, $ip = '0:0:0:0')
    {
        if ($shortcode == $this->config['short_code']) {
            $mg = new $this->mobileOriginated;
            $mg->msisdn = $msisdn;
            $mg->message = $message;
            $mg->transaction_id = $transactionId;
            $mg->received_at = Carbon::now();
            $mg->driver = $this->driverName;
            $mg->creator_ip = $ip;

            if ($mg->save()) {
                switch (strtolower(trim($message))) {
                    case '+':
                        $this->log('empty message');
                        $this->log('simple-sdp called mt (inform)');

                        $this->sendMt($msisdn, [
                            'content_id' => 0,
                            'message'    => trans("simple-sdp::messages.{$this->config['trans_prefix']}.inform")
                        ]);

                        break;

                    default:
                        $this->log('simple-sdp called mt (guide)');

                        $this->sendMt($msisdn, [
                            'content_id' => 0,
                            'message'    => trans("simple-sdp::messages.{$this->config['trans_prefix']}.guide")
                        ]);
                        break;

                }

                $this->log('done');
                return $this->response(true, Enum::SUCCESS_CODE);
            }
            else {
                $this->log('unknown error', 'error');
                return $this->response(false, Enum::UNKNOWN_ERROR_CODE);
            }
        }
        else {
            $this->log('not valid input (shortcode)', 'error');
            return $this->response(false, Enum::NOT_VALID_INPUTS_CODE);
        }
    }

    public function batchMo(Request $request)
    {
        $this->log('batch mo received');

        $xml = $request->getContent();

        $cleanXml = str_ireplace(['soapenv:', 'loc:'], '', $xml);
        $xml = simplexml_load_string($cleanXml);
        $collection = $xml->Body->sendSms->sendSmsCollection->SendSms;

        $status = 1;
        $result = [];

        foreach ($collection as $item) {
            $msisdn = (string)$item->addresses;
            $shortcode = (string)$item->senderName;
            $transactionId = (string)$item->correlator;
            $message = (string)$item->message;

            $response = $this->handleIncome($msisdn, $shortcode, $transactionId, $message, $request->ip());
            $result[] = $response;

            $status = $status * (int)$response->status;

        }

        if ($status) {
            $this->log('done');
            $this->log($result);

            return $this->response(true, Enum::SUCCESS_CODE, null, $result);
        }
        else {
            $this->log('failed', 'error');
            $this->log($result, 'error');

            return $this->response(false, Enum::UNKNOWN_ERROR_CODE, null, $result);
        }
    }
}